daemon/cmd/healthz.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"context"
+	"net"
+	"net/http"
+	"syscall"
+
+	"github.com/cilium/cilium/api/v1/models"
+
+	"golang.org/x/sys/unix"
+)
+
+func setsockoptReuseAddrAndPort(network, address string, c syscall.RawConn) error {
+	var soerr error
+	if err := c.Control(func(su uintptr) {
+		s := int(su)
+		// Allow reuse of recently-used addresses. This socket option is
+		// set by default on listeners in Go's net package, see
+		// net setDefaultListenerSockopts
+		soerr = unix.SetsockoptInt(s, unix.SOL_SOCKET, unix.SO_REUSEADDR, 1)
+		if soerr != nil {
+			return
+		}
+		// Allow reuse of recently-used ports. This gives the agent a
+		// better change to re-bind upon restarts.
+		soerr = unix.SetsockoptInt(s, unix.SOL_SOCKET, unix.SO_REUSEPORT, 1)
+	}); err != nil {
+		return err
+	}
+	return soerr
+}
+
+// startAgentHealthHTTPService registers a handler function for the /healthz
+// status HTTP endpoint exposed on addr. This endpoint reports the agent health
+// status and is equivalent to what the `cilium status --brief` CLI tool reports.
+func (d *Daemon) startAgentHealthHTTPService(addr string) {
+	lc := net.ListenConfig{Control: setsockoptReuseAddrAndPort}
+	ln, err := lc.Listen(context.Background(), "tcp", addr)
```

Comment:
I've never used a `ListenConfig` before and I couldn't find any documentation about the context. I'm wondering however if we should we cancel this context if `srv.Serve()` returns? In the end, I guess it doesn't matter that much, since it only returns on shutdown (or error, where we exit as well).

Reply:
Yeah, I guess it doesn't matter much in the end, but I could address it in a follow-up PR if need be.